feat: new DocTypes "Code List" and "Common Code"

barredterra opened the pull request:
![code_list](https://github.com/user-attachments/assets/db57cf07-4b84-487c-9bc2-25d606054fdc)

![common_code_list](https://github.com/user-attachments/assets/2196d3c6-0280-4797-9303-377c54c9e6f4)

![common_code](https://github.com/user-attachments/assets/ad2b534d-ce11-4fb0-bc1a-628b3a45bf9d)

```python
In [1]: code_list = "urn:xoev-de:kosit:codeliste:untdid.5305_3"

In [2]: get_codes_for(code_list, "Account", "Tax Assets - WPL")
Out[2]: ('A',)

In [3]: get_docnames_for(code_list, "Account", "A")
Out[3]: ('Tax Assets - WPL',)

In [4]: get_default_code(code_list)
Out[4]: 'S'
```

Screenshots based on https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.5305_3:technischerBestandteilGenericode:datei:UNTDID_5305-3.xml

After this is merged, check if we should revert https://github.com/frappe/erpnext/pull/40593 and https://github.com/frappe/erpnext/pull/40663

> no-docs (for now)

barredterra committed:
feat: new DocTypes "Code List" and "Common Code"

barredterra committed:
feat: add Common Code to Code List's connections

barredterra committed:
fix(Common Code): naming and unique constraint

barredterra committed:
feat(Common Code): add nicer, user-facing uniqueness validation

The error message from the DB unique contraint isn't very helpful.

blaggacao committed:
feat: Add index for code_list and common_code in CommonCode doctype

blaggacao committed:
refactor: create edi module

blaggacao committed:
feat: add genericode importers

blaggacao commented:
- [x] added fully tested genericode importer
- [x] added wip of an edi workspace
- [ ] make it hidden by default (check it actually works)
- [x] added wip of edi template
- [ ] dogfood and validate snippets with current (operational) implementation of colombia e-fact
- [ ] try out jpype and use java libraries for validation
- [ ] try to implement a wsdl endpoint per edi_template (hence `get_parsed`)

<details><summary>Wsdl service skeletton</summary>

```python
from spyne import Application, rpc, ServiceBase, Unicode
from spyne.protocol.soap import Soap11
from spyne.server.wsgi import WsgiApplication

class HelloWorldService(ServiceBase):
    @rpc(Unicode, _returns=Unicode)
    def say_hello(ctx, name):
        return f"Hello, {name}!"

application = Application([HelloWorldService],
    tns='http://example.com/hello',
    in_protocol=Soap11(validator='lxml'),
    out_protocol=Soap11())

# Generate WSDL
from spyne.interface.wsdl import Wsdl11
wsdl_content = Wsdl11(application.interface).get_interface_document()

# Save WSDL to file
with open('hello_service.wsdl', 'wb') as f:
    f.write(wsdl_content)
```

</details>

barredterra committed:
feat: validate distinct references

Ensure the same reference is not used on a different Common Code of the same Code List.

barredterra commented:
- [x] It is not obvious (and was not intended by me) that the `CanonicalVersionUri` now has to be used as the name of the code list (for the import to work). E.g. I would not expect the `CanonicalVersionUri` of the code list linked above to be referenced in most eInvoices. So a more generic name would be useful for retrieving the necessary code list. In the screenshot above, I used the `Identification/ShortName`.
- [x] "Additional Data" from the code list is not yet imported (e.g. descriptions, as available in the code list linked above).

blaggacao commented:
More exploration from my side: https://github.com/barredterra/erpnext/pull/1 (Edi Template & Edi Log, squashed your naming fix 
there, @barredterra )

blaggacao commented:
> It is not obvious (and was not intended by me) that the CanonicalVersionUri now has to be used as the name of the code list (for the import to work). E.g. I would not expect the CanonicalVersionUri of the code list linked above to be referenced in most eInvoices.

**Firstly**, most of the time, however, unlike the common codes, these lists are akin to real _constants_ in an e-invoice template. I would typically hard-code them for each distinct edi template. Here is an example:

<details><summary>XML Example</summary>

```xml
        <cbc:IdentificationCode
          listAgencyID="6"
          listAgencyName="United Nations Economic Commission for Europe"
          listSchemeURI="urn:oasis:names:specification:ubl:codelist:gc:CountryIdentificationCode-2.1"
        >CO</cbc:IdentificationCode>
```

</details>

**Secondly**, the title is still user friendly, for display purposes.

<details><summary>Screenshot</summary>

![image](https://github.com/user-attachments/assets/2bcf0ba3-5c25-4fe3-9b25-b7abeec74664)

</details>

**Thirdly**, as per the Genericode XML Schema, `CanonicalVersionUri` is the primary key of a list, so we would have to go out of our way to work around that. https://docs.oasis-open.org/codelist/genericode/v1.0/os/xsd/genericode.xsd

<details><summary>XML Reference</summary>

```xml
<xsd:complexType name="Identification">
<xsd:annotation>
<xsd:documentation>Identification and location information (metadata).</xsd:documentation>
</xsd:annotation>
<xsd:sequence>
<xsd:group ref="gc:NameSet">
...
</xsd:group>
<xsd:element name="Version" type="xsd:token">
...
</xsd:element>
<xsd:element name="CanonicalUri" type="xsd:anyURI">
<xsd:annotation>
<xsd:documentation>Canonical URI which uniquely identifies all versions (collectively).</xsd:documentation>
<xsd:documentation>
<rule:text id="R25" category="document">Must be an absolute URI, must not be relative.</rule:text>
</xsd:documentation>
<xsd:documentation>
<rule:text id="R26" category="application">Must not be used as a de facto location URI.</rule:text>
</xsd:documentation>
</xsd:annotation>
</xsd:element>
<xsd:group ref="gc:VersionLocationUriSet">
...
</xsd:group>
<xsd:element name="Agency" minOccurs="0" type="gc:Agency">
...
</xsd:element>
</xsd:sequence>
</xsd:complexType>

...

<xsd:group name="VersionLocationUriSet">
<xsd:annotation>
<xsd:documentation>Identification and location URIs for a version.</xsd:documentation>
</xsd:annotation>
<xsd:sequence>
<xsd:element name="CanonicalVersionUri" type="xsd:anyURI">
...
</xsd:element>
<xsd:element name="LocationUri" minOccurs="0" maxOccurs="unbounded" type="xsd:anyURI">
...
</xsd:element>
<xsd:element name="AlternateFormatLocationUri" minOccurs="0" maxOccurs="unbounded" type="gc:MimeTypedUri">
...
</xsd:element>
</xsd:sequence>
</xsd:group>
```

Note, that `CanonicalUri` _will_ be shared between lists, and within `VerstionLocationUriSet` (the unique identifying set), `CanoncialVersionUri` is the only mandatory one.

</details>

**Edit:**

I think this is what you where looking for:

![image](https://github.com/user-attachments/assets/dd4ac355-aaa2-4060-bc3a-7ec1421d864c)

blaggacao committed:
fix: additional data import

blaggacao committed:
style: semgrep

blaggacao committed:
fix: show title field in links

blaggacao reviewed (dismissed):
I have contributed the Genericode Importer which triggered a certain amount of polishing and I think this version is ready to merge, now.

We should keep in mind, that we're actively developing the EDI functionality for EU & Colombia in an iterative approach  against `develop` - so more to come, soon.

barredterra commented:
If we have to use this very technical name, we should at least offer the import from the list view (or from an unsaved form) instead of from the saved form view. It would be kind of weird if the user first needs to manually extract some values from the XML, in order to then import the rest of it.

My approach was originally targeted at, e.g., being able to tell the users to create a **Code List** called "UNTDID.5305" and make sure it contains the tax codes they care about. The alternative, equally good or even better, would be being able to tell them to just import [this code list](xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.5305_3:technischerBestandteilGenericode:datei:UNTDID_5305-3.xml). Telling them to extract the `CanonicalVersionUri` does not seem realistic, however.

blaggacao commented:
> [ ... ] we should at least offer the import from the list view (or from an unsaved form) instead of from the saved form view. It would be kind of weird if the user first needs to manually extract some values from the XML, in order to then import the rest of it.

> Telling them to extract the CanonicalVersionUri does not seem realistic, however.

Indeed! &mdash; I think this concern is already gracefully implemented?

### General Scenario: Add more codes from the same list (e.g. different filters) &mdash; form import
#### Specific Scenario: I accidentially selected the wrong code list for upload
![image](https://github.com/user-attachments/assets/513d647a-7935-43bd-8bba-66d6564ffb01)

---

### General Scenario: Add a new code list and some codes &mdash; list import
Implemented on the list views of _Code List_ and _Common Code_
#### Specific Scenario: Step Two (immediately after upload); Selection of Import filters
![image](https://github.com/user-attachments/assets/176017c5-5526-451e-9e05-130de1253c2f)
<sup>Note: no extraction of `CanonicalVersionUri` by the user; filterable are fields with cardinality <= 5</sup>

blaggacao committed:
fix: add agency id and fall back to longname on agency

blaggacao committed:
fix: gacefully handle parsing errors

blaggacao commented:
https://github.com/frappe/erpnext/pull/43425/commits/d540d03d8a5903764c311b8d19c3d7cb3188e6d9

![image](https://github.com/user-attachments/assets/e6d5caca-de0b-4373-bbe5-ce5a76454bbd)

blaggacao committed:
feat: add column selection heuristics

blaggacao committed:
feat: add optional code list URL

blaggacao committed:
fix: guard against undeclared columns

blaggacao committed:
fix: creation tracking

blaggacao commented:
We may need to relax the unique condition on `code` for _Common Code_. This is a valid Code List:
```xml
      <Row>
         <Value ColumnRef="code">
            <SimpleValue>1.50</SimpleValue>
         </Value>
         <Value ColumnRef="name">
            <SimpleValue>ReteFuente</SimpleValue>
         </Value>
         <Value ColumnRef="description">
            <SimpleValue>Compras con tarjeta débito o crédito</SimpleValue>
         </Value>
      </Row>
      <Row>
         <Value ColumnRef="code">
            <SimpleValue>1.50</SimpleValue>
         </Value>
         <Value ColumnRef="name">
            <SimpleValue>ReteFuente</SimpleValue>
         </Value>
         <Value ColumnRef="description">
            <SimpleValue>Compras de bienes o productos agrícolas o pecuarios sin procesamiento industrial</SimpleValue>
         </Value>
      </Row>
```

blaggacao committed:
fix: uniqueness on common_code does not hold true

blaggacao committed:
fix: add description to common code to help uniquely identify a code

blaggacao commented:
Problem addressed, also in the importer:

![image](https://github.com/user-attachments/assets/fa931e15-b20a-4252-a34e-bd5de3712baf)

![image](https://github.com/user-attachments/assets/e3f5e4f3-ea7c-4e35-89df-4f7b9b1bd99c)

blaggacao committed:
fix: make additional data read-only and avoid accidential schema invalidations

blaggacao committed:
fix: add code list identifier to the import modal

blaggacao commented:
https://github.com/frappe/erpnext/pull/43425/commits/52dcb2bca220aa4def00e5a338fb11b60943fe02 addresses the usability concern even further:

![image](https://github.com/user-attachments/assets/7aa2b042-fa1c-40cd-a734-bb74966add60)

blaggacao committed:
fix(edi-import): Add description column only when more than two columns are available

blaggacao commented:
## Testing Resume

- [x] 23 Code Lists, 27 XML files
- [x] Faulty XML: ok
- [x] Undeclared Columns: ok
- [x] Non unique `common_code`: ok (eg tax rates) -> added description
- [x] Guard against accidential structured data manipulation
- [x] Partial loading of a list
- [x] Later completion of a list - USER ATTENTION! Duplicates are legal in the data structure, double loading must be avoided with care (no other way)
- [x] Loading values from different files into the same code list 

<details><summary>Screenshot</summary>

![image](https://github.com/user-attachments/assets/436e9bba-b7bf-405c-a7b4-370df7b86e8a)

</details>

barredterra commented:
> Indeed! — I think this concern is already gracefully implemented?

Ah, my bad, I've overlooked this.

---

I think, from a developer perspective, it would be perfect if we can:

(1) Given a ref_doctype, ref_docname and code list, retrieve the applicable code.
(2) Given a code and code list, retrieve the applicable ref_doctype and ref_docname.

Do you think this unambiguous mapping is feasible?

I'm thinking of adding a util and validation for the second case.

---

Selecting "URL" on file upload fails. Would be nice if we can support this as well.

---

barredterra commented on erpnext/edi/doctype/code_list/code_list_import.py:
```diff
@@ -0,0 +1,104 @@
+	file_path = file_doc.get_full_path()
+
+	# Get available columns and example values
+	columns, example_values, filterable_columns = get_columns_and_examples(file_path)
```
```suggestion
	columns, example_values, filterable_columns = get_columns_and_examples(root)
```

I think we can pass the already parsed XML here.

blaggacao committed:
chore: pass already parsed root

blaggacao commented:
> (2) Given a code and code list, retrieve the applicable ref_doctype and ref_docname.

<details><summary>Given ...</summary>

```python

	def validate_distinct_references(self):
		"""Ensure no two Common Codes of the same Code List are linked to the same document."""
		for link in self.applies_to:
			existing_links = frappe.get_all(
				"Common Code",
				filters=[
					["name", "!=", self.name],
					["code_list", "=", self.code_list],
					["Dynamic Link", "link_doctype", "=", link.link_doctype],
					["Dynamic Link", "link_name", "=", link.link_name],
				],
				fields=["name", "common_code"],
			)

			if existing_links:
				existing_link = existing_links[0]
				frappe.throw(
					_("{0} {1} is already linked to Common Code {2}.").format(
						link.link_doctype,
						link.link_name,
						get_link_to_form("Common Code", existing_link["name"], existing_link["common_code"]),
					)
				)
```

</details>

I think we can restrict certain Code Lists (optionally) to a certain doctype and thereby meet the full condition necessary to implement such a util for such code lists.

But you're absolutely right: on the parsing end of the flow, this would quite definitely improve DX.

blaggacao committed:
fix(edi-import): support link & library files, too

blaggacao commented:
> Selecting "URL" on file upload fails. Would be nice if we can support this as well.

https://github.com/frappe/erpnext/pull/43425/commits/94bd74319ba8f5641c1ff5dc5e192fe296bd0113

Also library links (for example to post-load more codes from an already partially imported list)

Depends on: https://github.com/frappe/frappe/pull/27954

blaggacao commented:
We probably should also return `additional_data` from the `get_code_for` because more often than not, this is actually relevant data when rendering a e-invoice template. I'll check this and confirm during dogfooding.

---

Wile dogfooding I realise:

As user, I want an easy way to associate all Code List documents with my current document, which are configured to be associatable with that document type (by means of a Code List - Doctype association). Especially for documents which are still metadata, but fall within the life-cycle management of non-power users, such as Suppliers, Customers & Items.

---

We also should probably implement a stringer (`__str__`) on `CommonCode` which automatically renders it's `common_code` into a jinja template without need for more verbosity.

---

`get_code_for` should probably throw a user facing error if no association is found but was expected, indicating that xyz association is missing and needs to be set up.

```python
    # should throw if tax_category_id was not set
    tax_category_id = frappe.get_doc("Code List", TAX_CATEGORY_CODE_LIST).get_code_for(
        party.doctype, party.name
    )
```

---

_Code List_ should have a default code, so that it is easier to fetch:

```python
    tax_id_type = frappe.get_doc("Common Code", party.tax_id_type or tax_type_codelist.default)
```

---

Would be handy too (`.get_doc` proxy):
```python
    tax_id_type = codelist.get_doc(party.tax_id_type)  # raises
    tax_id_type = codelist.get_doc(party.tax_id_type, None)
    tax_id_type = codelist.get_doc(party.tax_id_type, codelist.default)
```

---

Ultimately, what I want is this, to keep my implementation at the leaf as clean as possible:
```python
from .code_list import substitute


customer = frappe.get_doc("Customer", "Foo")

customer_for_render = substitute(customer)
# replaces all links to Common Code to the respective common code
# and infers defaults based on the field's filter
# {"code_list": "My List"} - take default of that list, if there is
```

<sup> If my edge case invalidates a default, I can still override it in sitiu </sup>

blaggacao commented:
@barredterra I noted a couple of things, when dogfooding this yesterday.

1. On "intermediate" metadata (like Customer, Supplier, which are metadata, but where the lifecycle is managed as part of daily transactions), I think we still want a "nomarl" link from a custom field to Common Code with the appropriate filter.
2. On these cases it would be wonderful to have a summary `customer = code_list.substitute(customer)` to replace all these link fields with a Common Code object for use during rendering. (benefits: same mental model as the origin doctype, just augmented, avoids running into the trap of creating own `frappe._dict()` ad-hoc data structures)
3. Because then, we get into Custom Field management already, I think it would be fair to add a (rudimentary) wizard of some sort to the Code List doctype which allows to setup such custom fields &mdash; this would be an alternative mechanism of association for the use case of those "indermediate" metadata in addition to the Common Code's `applies_to` for "long lived" metadata.

This doesn't mean the `applies_to` is useless, it's still extremely valid on quite an array of very stable metadata, but I think for the transaction-bound metadata (read: Customer), we need an in-transaction complement.

You may consider my previous list as "dogfooding notes", I haven't fully processed them myself, yet. But it seems that my refactoring would be stuck on what I describe here, for the "true blessing of DX".

blaggacao commented:
> (2) Given a code and code list, retrieve the applicable ref_doctype and ref_docname.

Since this happens in the context of parsing, and parsing is bound to a XML structure, the following binding may be relevant, as well:

![image](https://github.com/user-attachments/assets/e4bc9697-77ea-4086-9fa7-17555e5e7dc7)

**Meaning**
- In the context of DIAN E-Invoice
- With its binding to a Sales Invoice
- And the associated code lists bound to python context variables `pais` & `to` ("given code list")
- Then infer the code's `refdoc_doctype` & `refdoc_name` by means of a triangulation with an additional Doctype Context binding in the `applies_to` table.

Originally, I had thought of an implementation with a direct Code List - Doctype binding, but I realize that the EDI Template is a better source for that contextual binding. Hence having such a context binding within the `applies_to` table might make more sense.

It would read in plain english:

1. In the context of Sales Invoice EDI
2. Apply the present code
3. To the target document (e.g. an _Account_ / Tax) 

Now, instead of binding it to a Doctype, we probably could bind it to an EDI Template within `applies_to` to make true with the above plain english reading of the feature.

blaggacao commented:
... which makes me wonder: probably we should bind all `applies_to` inside the EDI template (as in: "Within the context of a particular EDI type")? :thinking: 

---

![image](https://github.com/user-attachments/assets/1b3b4a83-a18a-49f1-adf7-e2524b7bc19c)

blaggacao committed:
fix(edi-import): require title on common code

blaggacao commented:
https://github.com/frappe/erpnext/pull/43425/commits/d68d80892db14b62e47f60a22b441fa62e5b77b6

Fix UX of Title-less common codes in link dropdowns:

![image](https://github.com/user-attachments/assets/64fc084f-90d1-460a-8be8-e984e71514e2)

blaggacao commented:
Conclusion: We need to incubate a core feature "LazyRecords", which lazily traverse link values on on access but eagerly build the traversal tree 3-4 levels deep for best autocompletion. Inside, we then can replace standard Link bindings with our custom bindings. Whereby barren key renders the stringer of the target class in jinja (typically: name or title(?)). :rofl:

barredterra committed:
fix: don't nest import button

We don't have multiple imports yet, so we can save one click and improve dicoverability.

barredterra committed:
refactor: import genericode

- Allow re-import of the same code list to update existing data
- Fill all Code List data on first creation, not just the name
- Move import of Common Codes to common_code.py
- Add type hints
- Fix naming settings
- Move static utils out of class
- Remove unused import
- Remove ignore_permission=True

barredterra committed:
feat: show code list title and link in import modal

barredterra committed:
refactor(Common Code): get_code_for

barredterra committed:
feat(Common Code): util `get_doc_for`

barredterra commented:
https://github.com/frappe/erpnext/pull/43425#issuecomment-2385467354

I broke this again, sorry.

Is this really valid? What use is a code if it can have multiple meanings, even in the same list?

blaggacao commented:
> Is this really valid?

Yes.

> What use is a code if it can have multiple meanings, even in the same list?

Not sure if we productively can second-guess the spec &mdash; especially since I have a real codelist which exhibits this behaviour.

Remember that `code` is not guaranteed to be a row prim key (there isn't even a guarantee that a row prim key is present), so we can't fake it and make it a requirement in our implementation.

blaggacao commented on erpnext/edi/doctype/common_code/common_code.py:
```diff
@@ -60,52 +59,68 @@ def validate_distinct_references(self):
+			common_code.code_list = code_list
+
+		common_code.from_genericode(column_map, xml_element)
+		common_code.save()
```
This is unfortunately prohibitively slow :cry: &mdash; we should revert back to the bulk insert.

I don't quite see a use case to update a code list item: they don't change on the same version. If they change they change on a different version which is a different code list.

barredterra replied:
Prohibitively slow? I could't even blink while this imported my test data. How big are your code lists and how many do you want to import?

To be a bit more precise, on my machine this has a throughput of >200 records per second. And it needs to be done only once.

blaggacao replied:
> Prohibitively slow?

It was at least pretty annoying during local tests, so much so that I took care to make it faster (not by accident or not knowing better). Now you took the same care to make it slower (for other reasons not entirely evident to even though I tried to infer some), so I'll have to assume it's going to be "pretty annoying during local tests" again.

Maybe you want to first briefly explain these "other reasons"?

barredterra replied:
I usually avoid `bulk_insert`. Yes, it is fast, but the reason for that is that it skips all validations and hooks, making it error prone and hard to customize. I would rather wait a couple seconds instead of skipping all that good stuff. Sure, when a task would take many minutes or hours, we cannot avoid bulk_insert. But for saving a couple of seconds, I think it's not worth the (potential) trouble.

barredterra replied:
I have now added a progress bar, so we don't have to stare at a blank screen during longer imports.

blaggacao replied:
I see. Thanks. That makes sense, in principle.

Do we realistically expect the need for validations and hooks on _mass_ import of _Common Codes_ in practice, though? &mdash; I have a bit of a hard time to come to terms with the ask and presumption to deprioritize speed on abstract principle, alone.

The scenario here simply seems a bit far fetched to me.

The progress bar is a good idea to mitigate the problem. At least people wouldn't try to reload the page because it goes unresponsive.

It's been a while and I can't remember exactly how long it took. Granted, "prohibitively", indeed says as much about my personal tolerance level for slowness. 😊 Let me check it again report back, and when it takes more than 5 seconds, we revert back?

blaggacao replied:
Checking now:

![image](https://github.com/user-attachments/assets/bcd297a9-415c-410e-926f-2978fc1a14fb)

blaggacao replied:
Freeze screen for about 30 sec:


![image](https://github.com/user-attachments/assets/7135cdba-76f1-4365-a99c-f1ceeebd01e9)



That's not instant (as before), but probably ok.

blaggacao replied:
Loading, otoh, was 8 sec. That's totally fine.

barredterra replied:
> Freeze screen for about 30 sec:
> That's not instant (as before), but probably ok.

iirc, deletion was always slow. didn't change anything here.

blaggacao commented on erpnext/edi/doctype/common_code/common_code.py:
```diff
@@ -60,52 +59,68 @@ def validate_distinct_references(self):
+	for xml_element in root.xpath(xpath_expr):
+		code_column = column_map["code"]
+		code_value = xml_element.find(f"./Value[@ColumnRef='{code_column}']/SimpleValue").text
+		code_name = get_code_name(code_list, code_value)
```
`idx` is required here to prevent name conflicts (make them less likely \*). The hashing is still valuable as it provides predictable / independently reproducible names.

\* They can still occur when importing from two files where exactly at the same `idx` (after filtering) we have the same `code` &mdash; we should fix that better, actually.

barredterra replied:
So let's just use normal hash naming? I don't see the need for predictable / independently reproducible names and using idx seems kind of random (except the order is guaranteed within the code list).

blaggacao replied:
> So let's just use normal hash naming?

Yeah, that would solve the issue for good. :+1:

blaggacao reviewed (dismissed):
I appreciate the better codelist handling &mdash; that looks a lot cleaner.

We'd just need to mitigate the two regressions:

- Insertion speed for Codes (removal of raw bulk insert) &mdash; I think this can be still achived after the clarifying refactorings by operating on a generator / list rather than on the individual record object
- Unique naming &mdash; we should aim to fix the silent bug in the previous implementation, as well

> I broke this again, sorry.

Maybe I overlooked. But I'm not even quite sure if you broke it in principle :smile: or just accidentially via removing the `idx` during import. :smile:

blaggacao commented on erpnext/edi/doctype/code_list/code_list.py:
```diff
@@ -78,3 +82,23 @@ def get_code_for(code_list: str, doctype: str, name: str) -> str | None:
 	return codes[0][0] if codes else None
+
+
+def get_doc_for(code_list: str, doctype: str, code: str) -> str | None:
+	"""Return the record name for a given common code"""
+	CommonCode = frappe.qb.DocType("Common Code")
+	DynamicLink = frappe.qb.DocType("Dynamic Link")
+
+	docnames = (
+		frappe.qb.from_(CommonCode)
+		.join(DynamicLink)
+		.on((CommonCode.name == DynamicLink.parent) & (DynamicLink.parenttype == "Common Code"))
+		.select(DynamicLink.link_name)
+		.where(
+			(DynamicLink.link_doctype == doctype)
+			& (CommonCode.common_code == code)
+			& (CommonCode.code_list == code_list)
+		)
+	).run()
+
+	return docnames[0][0] if docnames else None
```
So, I guess this will break in a subset of valid cases, namely when we can't assume `common_code` to be a unique column, right?

I think we should avoid to offer unbound behaviour. For example, we could return a UserError condition if we detect more than one row in the query result. This would even relax the unique constraint required by this getter to `(codelist, doctype, code)` without precluding that the underlying dataset can't exhibit such records.

barredterra committed:
fix: helper methods cannot assume unique records

Return a tuple of results instead

barredterra committed:
fix(Common Code): switch to hash naming

barredterra committed:
fix(Common Code): support empty columns in imported file

Value can be a self-closing tag without a nested SimpleValue, i.e. `<Value ColumnRef="Alternative-Einheit"/>`

barredterra committed:
feat: show progress bar during code list import

barredterra committed:
refactor: remove unused return value, simplify structure

barredterra commented:
For testing, here's a list of code lists required for eInvoicing in Germany (XRechnung):

- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:rec21_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:rec20_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:vatex_1:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:eas_5:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:icd_5:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.7161_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.7143_4:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.5305_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.5189_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:xrechnung:codeliste:untdid.4461_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.4451_4:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.2005_4:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.1153_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:untdid.1001_4:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:currency-codes_3:technischerBestandteilGenericode
- https://www.xrepository.de/api/xrepository/urn:xoev-de:kosit:codeliste:country-codes_8:technischerBestandteilGenericode

barredterra committed:
fix: cut off excessively long titles

barredterra committed:
feat: more default values in import dialog

barredterra commented on erpnext/edi/doctype/common_code/common_code.py:
```diff
@@ -0,0 +1,118 @@
+	def validate(self):
+		self.validate_distinct_references()
+
+	def validate_distinct_references(self):
+		"""Ensure no two Common Codes of the same Code List are linked to the same document."""
```
For generating messages, it is very useful to have exactly one common code per ERPNext entity. However, for parsing messages, it would be useful to map multiple common codes to the same ERPNext entity. This is currently not possible due to this validation.

barredterra replied:
@blaggacao any idea how to resolve this?

blaggacao replied:
I'd recommend binding in the context of an EDI Template. Within this narrowed down context, this restriction is likely more suitable, but also the binding list is usually short enough that no duplicates would typically happen.

But this restriction, regardless, makes sense `on_update` within https://github.com/frappe/erpnext/pull/44174, specifically on [`bound_common_codes`](https://github.com/frappe/erpnext/pull/44174/files#diff-f3b34a58e2b8bf8d6892dfd41d9a5a4d076535000ad54d7ef12f669326128cd5R46)

blaggacao commented on erpnext/edi/doctype/common_code/common_code.py:
```diff
@@ -28,6 +28,10 @@ class CommonCode(Document):
 		title: DF.Data
 	# end: auto-generated types
 
+	def before_validate(self):
+		if self.title and len(self.title) > 140:
+			self.title = self.title[:139] + "…"
+
```
If the title is used as canonical reference value (which I know I do), this might just start **randomly** braking some XML validations.

I'd construe this to be a view, not a data issue.

barredterra replied:
What's the alternative? Our `Data` fieldtype doesn't support arbitrarily long data.

blaggacao replied:
We'd have essentially two alternatives:

- Let it fail an the current field length (eventually with a better error message)
- Let it fail (very much less likely so) on an increased field length (eventually with a better error message), e.g. of `length: 300`

> Our `Data` fieldtype doesn't support arbitrarily long data.

We can dial it up, though.

blaggacao replied:
We could instruct the user tho dial up the field length on failure via customization, for example.

barredterra replied:
Increased the length, will leave the error message as is.

blaggacao reviewed (changes requested):
I don't think manipulating origin reference data is a good idea. This may seem like a view, not a data problem.

barredterra committed:
Revert "fix: cut off excessively long titles"

This reverts commit 848cf041e70021df38beeb8c3d1db6166846f171.

barredterra committed:
fix(Common Code): increase title length to 300

barredterra committed:
perf(Common Code): index code_list and common_code

barredterra committed:
feat(Common Code): set search fields

blaggacao reviewed (commented):
Dismiss prior review :smile: I just need to test this a little more as a dep for `edi_template.py` and Colombian E-Invoice implementation. Getting there :hourglass_flowing_sand: 

Hm, that didn't work.... :thinking:

barredterra committed:
feat(Code List): default common code

Allow linking to a Common Code that shall be used as the default. Add helper methods to retrieve the default code.

barredterra commented on erpnext/edi/doctype/code_list/code_list.py:
```diff
@@ -0,0 +1,125 @@
+		)
+
+		for doc in linked_docs:
+			frappe.delete_doc("Common Code", doc.name, force=1)
```
```suggestion
			frappe.delete_doc("Common Code", doc.name)
```

@blaggacao was this intentional? If there are any existing links to such a force-deleted Common Code, the linking form will break.

blaggacao replied:
Hm, this looks like a dev-time QoL feature, in hindsight.

Deletion should almost never happen in normal operation.

blaggacao commented:
Rebase done in order to prepare internal production backport which is hard to extract with a merge commit from develop.

barredterra committed:
fix:  don't force-delete

barredterra committed:
feat(Code List): add description to search fields

Description is the field that usually contains words that humans can make sense of. Show this to help select the correct list.

barredterra committed:
fix: sort by idx when retrieving docname

This way we can link multiple docs to one Common Code and sort them by preference.

blaggacao committed:
fix: import modal only on existing code list

blaggacao reviewed (approved):
LGTM, I'm using this via branch https://github.com/barredterra/erpnext/pull/1 and it seems stable enough for this use case

Some of [these notes](https://github.com/frappe/erpnext/pull/43425#issuecomment-2386141316) might or might not serve for future evolution.

blaggacao commented:
@barredterra This would need backports of https://github.com/frappe/frappe/pull/27954
